### PR TITLE
Vector transformation

### PR DESCRIPTION
**Pull request by junpenglao**
A super class wrapper for vector transformation.
Example usage:
```python
class Nealfun(VectorTransform):
    name = "Neal_funnel"

    def backward(self, y):
        x = tt.zeros(y.shape)
        x = tt.inc_subtensor(x[0], y[0] / 3.)
        x = tt.inc_subtensor(x[1:], y[1:] / tt.exp(y[0] / 2))
        return x

    def forward(self, x):
        y = tt.zeros(x.shape)
        y = tt.inc_subtensor(y[0], x[0] * 3.)
        y = tt.inc_subtensor(y[1:], tt.exp(x[0] * 3. / 2) * x[1:])
        return y
```

### CODE REVIEW

File: pymc3/distributions/transforms.py [hunk 1]
```diff
@@ -50,6 +50,16 @@ def jacobian_det(self, x):
         return tt.log(tt.abs_(grad))
 
 
+class VectorTransform(Transform):
+
```
Comment (springcoil) [inline pymc3/distributions/transforms.py:54]: Maybe a quick docstring here would be useful

**Comment by springcoil** [comment]
So cool @junpenglao and I know it's a WIP. 

- tests (not sure off the top of my head what tests you'd do, maybe just confirming your example usage does what you want) 
- docs with example usage maybe in http://docs.pymc.io/advanced_theano.html

**Comment by junpenglao** [comment]
Thanks for the comment @springcoil! I am thinking this might come in handy when we are doing coordinate transformation but I am not sure what is the right use case / test case yet.

Also, it might be easier to call the `jacobian` from `pymc3.theanof`.

**Comment by junpenglao** [comment]
This will be good for implementing Copulas - working on an example similar to https://github.com/tensorflow/probability/blob/master/tensorflow_probability/examples/jupyter_notebooks/Gaussian_Copula.ipynb

**Comment by springcoil** [comment]
Or https://twiecki.github.io/blog/2018/05/03/copulas/

**Comment by twiecki** [comment]
Your example is quite neat. I suppose you could pass in an `MvNormal` and it would transform it to a funnel?

**Comment by junpenglao** [comment]
Yep, I was using it as an example for my workshop, to show that if you account for the Jacobian it will cancel out the benefit of reparameterization (see the last part of https://github.com/junpenglao/Planet_Sakaar_Data_Science/blob/master/Miscellaneous/Neals_funnel.ipynb)

**Comment by springcoil** [comment]
Any update on this?

**Comment by junpenglao** [comment]
Thanks for the reminder - I still having find a good user case, not sure if this is really useful.

**Comment by springcoil** [comment]
Didn't you use it for copulas?

File: pymc3/distributions/transforms.py [hunk 2]
```diff
@@ -50,6 +50,16 @@ def jacobian_det(self, x):
...
+    def jacobian_det(self, x):
+        f_inv = self.backward(x)
+        J, _ = theano.scan(lambda i, f, x: tt.grad(f[i], x),
```
Comment (ferrine) [inline pymc3/distributions/transforms.py:57]: why hardcode tt.grad function? We may know closed form solution without scan

Comment (ferrine) [reply pymc3/distributions/transforms.py:57]: Interesting insight is that naive theano.gradient.jacobian works a bit slower than this implementation

**Comment by twiecki** [comment]
Closing due to inactivity.

**Pull request closed without merging**